File: examples/catalog/sku.go
package catalog

import (
	"fmt"
	"regexp"
)

var skuPattern = regexp.MustCompile(`^[A-Z0-9-]{3,20}$`)

type SKU struct {
	value string
}

func NewSKU(value string) (SKU, error) {
	if !skuPattern.MatchString(value) {
		return SKU{}, fmt.Errorf("invalid SKU %q: must be 3-20 characters of uppercase letters, digits, and hyphens", value)
	}
	return SKU{value: value}, nil
}

func MustNewSKU(value string) SKU {
	s, err := NewSKU(value)
	if err != nil {
		panic(err)
	}
	return s
}

func (s SKU) String() string {
	return s.value
}

File: examples/python/catalog/sku.py
import re
from dataclasses import dataclass

_SKU_PATTERN = re.compile(r"^[A-Z0-9-]{3,20}$")


@dataclass(frozen=True)
class SKU:

    _value: str

    def __post_init__(self) -> None:
        if not _SKU_PATTERN.match(self._value):
            raise ValueError(
                f"invalid SKU {self._value!r}: must be 3-20 characters of "
                "uppercase letters, digits, and hyphens"
            )

    def __str__(self) -> str:
        return self._value
